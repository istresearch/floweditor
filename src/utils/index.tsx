import {
  Action,
  Case,
  Category,
  ContactProperties,
  FlowPosition,
  LocalizationMap
} from 'flowTypes';
import { Query } from 'immutability-helper';
import * as React from 'react';
import Localization, { LocalizedObject } from 'services/Localization';
import { Asset, Assets } from 'store/flowContext';
import { FormEntry } from 'store/nodeEditor';
import { v4 as generateUUID } from 'uuid';
import variables from 'variables.module.scss';

export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
const LABEL_CHARS = /^[a-zA-Z0-9-][a-zA-Z0-9- ]*$/;
const MAX_LABEL_LENGTH = 36;
const SNAKED_CHARS = /\s|-+(?=\S)/g;

export const MAX_REFLOW_ATTEMPTS = 100;
export const ACTIVITY_INTERVAL = 5000;
export const COLLISION_FUDGE = 5;
export const GRID_SIZE: number = parseInt(variables.grid_size, 10) || 20;
export const NODE_SPACING: number = parseInt(variables.node_spacing, 10) || 10;
export const NODE_PADDING: number = parseInt(variables.node_padding, 10) || 10;
export const QUIET_NOTE = 2000;
export const CONFIRMATION_TIME = 2000;

export interface BoolMap {
  [key: string]: boolean;
}

interface Bounds {
  left: number;
  top: number;
  right: number;
  bottom: number;
}

export type LabelIdCb = (label?: string, labelKey?: string, valueKey?: string) => string;

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
  let leftAdjust = left % GRID_SIZE;
  let topAdjust = top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  return {
    left: Math.max(left + leftAdjust, 0),
    top: Math.max(top + topAdjust, 0)
  };
};

/**
 * Adjusts the position offsets to a grid
 */
export const snapPositionToGrid = (position: FlowPosition): FlowPosition => {
  let leftAdjust = position.left % GRID_SIZE;
  let topAdjust = position.top % GRID_SIZE;

  if (leftAdjust > GRID_SIZE / 3) {
    leftAdjust = GRID_SIZE - leftAdjust;
  } else {
    leftAdjust = leftAdjust * -1;
  }

  if (topAdjust > GRID_SIZE / 3) {
    topAdjust = GRID_SIZE - topAdjust;
  } else {
    topAdjust = topAdjust * -1;
  }

  const left = Math.max(position.left + leftAdjust, 0);
  const top = Math.max(position.top + topAdjust, 0);
  const right = left + position.right - position.left;
  const bottom = top + position.bottom - position.top;

  return {
    left,
    top,
    right,
    bottom
  };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
  array.reduce(
    (map: BoolMap, item: string) => ({
      ...map,
      [item]: true
    }),
    {}
  );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
  value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
  value
    .toLowerCase()
    .trim()
    .replace(SNAKED_CHARS, '_');
/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
  str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClassForEntry = (entry: FormEntry): string => {
  if (entry && entry.validationFailures && entry.validationFailures.length > 0) {
    return 'select-invalid';
  }
  return '';
};

export const getSelectClass = (errors: number): string => {
  if (errors === 0) {
    return 'react-select select-base';
  }
  // We use a global selector here for react-select
  return 'react-select select-base select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
  const [removed] = list.splice(startIndex, 1);

  list.splice(endIndex, 0, removed);

  return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
  JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
  if (!errors.length) {
    return false;
  }
  for (const error of errors) {
    for (const exp of exps) {
      if (error.match(exp)) {
        return true;
      }
    }
  }
  return false;
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
  if (localizations && localizations.length) {
    return localizations[0];
  }
};

export interface ClickHandler {
  // onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseUp: (event: React.MouseEvent<HTMLDivElement>) => void;
}

/**
 * Creates a simple click handler via onMouseDown and onMouseUp.
 * This is a necessity in order to let jsPlumb manage our element dragging.
 * @param onClick
 */
export const createClickHandler = (
  onClick: (event: React.MouseEvent<HTMLElement>) => void,
  shouldCancelClick: () => boolean = null,
  onMouseDown?: (event: React.MouseEvent<HTMLElement>) => void
): any => {
  return {
    onMouseDown: (event: React.MouseEvent<HTMLElement>) => {
      if (onMouseDown) {
        onMouseDown(event);
      }
    },
    onMouseUp: (event: React.MouseEvent<HTMLElement>) => {
      if (!shouldCancelClick || !shouldCancelClick()) {
        onClick(event);
      }
    }
  };
};

export const getLocalization = (
  obj: Action | Category | Case,
  localization: LocalizationMap,
  language: Asset
) => Localization.translate(obj, language, localization[language.id]);

/** istanbul ignore next */
export const dump = (thing: any) => console.log(JSON.stringify(thing, null, 4));

/**
 * Apply emphasis style
 */
export const emphasize = (text: string) => <span className="emphasize">{text}</span>;
export const ellipsize = (text: string, max: number = 250) => {
  if (text.length > max) {
    return text.substr(0, max) + '...';
  }
  return text;
};

/**
 * Does property exist in our ContactProperties enum?
 * @param {string} propertyToCheck - 'name' key on react-select AutoCompleteOption
 */
export const propertyExists = (propertyToCheck: string) => {
  for (const property of Object.keys(ContactProperties)) {
    if (property.toLowerCase() === propertyToCheck.toLowerCase().trim()) {
      return true;
    }
  }
  return false;
};

/**
 * Should x element be rendered?
 */
export const renderIf = (predicate: boolean) => (then: JSX.Element, otherwise?: JSX.Element) =>
  predicate ? then : otherwise ? otherwise : null;

/**
 * Does the label meet our length requirements?
 * @param {string} label - label created by react-select
 */
export const properLabelLength = (label: string = '') =>
  label.length > 0 && label.length <= MAX_LABEL_LENGTH;

/**
 * Does the label meet our character requirements?
 * @param {string} label
 */
export const containsOnlyLabelChars = (label: string = '') => LABEL_CHARS.test(label);

/**
 * Does the label meet our length, character requirements?
 * @param {string} label - label created by react-select
 */
export const isValidLabel = (label: string) =>
  properLabelLength(label) && containsOnlyLabelChars(label);

export const isRealValue = (obj: any) => obj !== null && obj !== undefined;

/* istanbul ignore next */
export const timeStart = (name: string) =>
  process.env.NODE_ENV === 'development' && console.time(name);

/* istanbul ignore next */
export const timeEnd = (name: string) =>
  process.env.NODE_ENV === 'development' && console.timeEnd(name);

export const log = (...message: any[]) =>
  process.env.NODE_ENV === 'development' && console.log(...message);

export const capitalize = (str: string) =>
  str.replace(/(?:^|\s)\S/g, captured => captured.toUpperCase());

export const set = (val: any): Query<any> => ({ $set: val });

export const setTrue = (): Query<true> => set(true);

export const setFalse = (): Query<false> => set(false);

export const setNull = (): Query<null> => set(null);

export const setEmpty = (): Query<string> => set('');

export const merge = (val: any): Query<any> => ({ $merge: val });

export const unset = (val: any): Query<any> => ({ $unset: val });

export const push = (arr: any[]): Query<any[]> => ({ $push: arr });

// tslint:disable-next-line:array-type
export const splice = (arr: Array<Array<any>>): Query<Array<Array<any>>> => ({
  $splice: arr
});

export const optionExists = (newOptName: string, options: any[]) =>
  options.find(({ name }) => name.toLowerCase().trim() === newOptName.toLowerCase().trim())
    ? true
    : false;

/* export const isOptionUnique: IsOptionUniqueHandler = ({ option, options, labelKey, valueKey }) =>
    !propertyExists(option.name) && !optionExists(option.name, options);

export const isValidNewOption: IsValidNewOptionHandler = ({ label }) => isValidLabel(label);
*/
/* export const composeCreateNewOption = ({
    idCb,
    type
}: {
    idCb: LabelIdCb;
    type: AssetType;
}): NewOptionCreatorHandler => ({ label, labelKey, valueKey }) => ({
    id: idCb(label, labelKey, valueKey),
    name: label,
    type,
    isNew: true
});*/

/**
 * Deduplicate values in an array.
 * Pass it a key to deduplicate based on obj key.
 */
export const uniqueBy = (a: any[], key: string): any[] => {
  const seen: any = {};
  return a.filter((item: any) => {
    const k = item[key];
    return seen.hasOwnProperty(k) ? false : (seen[k] = true);
  });
};

export const downloadJSON = (obj: any, name: string): void => {
  const dataStr =
    'data:text/json;charset=utf-8,' + encodeURIComponent(JSON.stringify(obj, null, 2));
  const downloadAnchorNode = document.createElement('a');
  downloadAnchorNode.setAttribute('href', dataStr);
  downloadAnchorNode.setAttribute('download', name + '.json');
  document.body.appendChild(downloadAnchorNode); // required for firefox
  downloadAnchorNode.click();
  downloadAnchorNode.remove();
};

export const fetchAsset = (asset: Asset) => (assets: Assets, id: string): Promise<Asset> => {
  return new Promise<Asset>((resolve, reject) => {
    resolve(asset);
  });
};

export let createUUID = (): string => {
  return generateUUID();
};

export const seededUUIDs = (seed: number = 1): any => {
  let current = seed;
  const randomNext = () => {
    const x = Math.sin(current++) * 10000;
    return x - Math.floor(x);
  };

  const random = (min: number, max: number): number => {
    return Math.floor(randomNext() * (max - min) + min);
  };

  return () => {
    const randomSeed: number[] = [];
    for (let idx = 0; idx < 16; idx++) {
      randomSeed.push(random(0, 250));
    }
    return generateUUID({ random: randomSeed });
  };
};

export const range = (start: number, end: number) =>
  Array.from({ length: end - start }, (v: number, k: number) => k + start);

export const pluralize = (count: number, noun: string, suffix: string = 's'): string =>
  `${noun}${count !== 1 ? suffix : ''}`;

export const hasString = (names: string[], key: string): boolean =>
  !!names.find((item: string) => item === key);

export const scalarArrayEquals = (a: any[], b: any[]) =>
  a.every((value, index) => value === b[index]);

export const getURNPath = (urn: string) => {
  return urn.split(':')[1];
};

export const copyToClipboard = (text: string) => {
  navigator.clipboard.writeText(text);
};

export const throttle = (func: any, timeout: any) => {
  let ready: boolean = true;

  return (...args: any) => {
    if (!ready) {
      return;
    }

    ready = false;
    func(...args);
    setTimeout(() => {
      ready = true;
    }, timeout);
  };
};

export const traceUpdate = (component: any, prevProps: any, prevState?: any) => {
  const messages: string[] = [];
  Object.entries(component.props).forEach(
    ([key, val]) => prevProps[key] !== val && messages.push(`Prop: '${key}' changed`)
  );
  if (prevState && component.state) {
    Object.entries(component.state).forEach(
      ([key, val]) => prevState[key] !== val && messages.push(`State: '${key}' changed`)
    );
  }

  if (messages.length > 0) {
    log('****  ' + component._reactInternalFiber.type.name + ' changed');
    messages.forEach((message: string) => {
      log(message);
    });
  }
};

export const debounce = (fn: any, quiet: number, closure: any = null) => {
  if (fn.timeout) {
    window.clearTimeout(fn.timeout);
  }
  fn.timeout = window.setTimeout(closure || fn, quiet);
};

export const onNextRender = (fn: any) => {
  window.setTimeout(fn, 0);
};

export const desnake = (text: string): string => {
  if (!text) {
    return text;
  }
  return text
    .split('_')
    .join(' ')
    .toLowerCase();
};

export const bool = (prop: boolean) => {
  return prop ? 'true' : null;
};
