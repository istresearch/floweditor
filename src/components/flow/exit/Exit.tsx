import { react as bindCallbacks } from 'auto-bind';
import classNames from 'classnames/bind';
import Counter from 'components/counter/Counter';
import DragHelper from 'components/draghelper/DragHelper';
import { getExitActivityKey } from 'components/flow/exit/helpers';
import Loading from 'components/loading/Loading';
import { fakePropType } from 'config/ConfigProvider';
import { Cancel, getRecentMessages } from 'external';
import { Category, Exit, FlowNode, LocalizationMap } from 'flowTypes';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { RecentMessage } from 'store/editor';
import { Asset } from 'store/flowContext';
import AppState from 'store/state';
import { DisconnectExit, disconnectExit, DispatchWithState } from 'store/thunks';
import { createClickHandler, getLocalization, renderIf } from 'utils';

import * as moment from 'moment';
import styles from './Exit.module.scss';
import { Portal } from 'components/Portal';
import i18n from 'config/i18n';

export interface RenderCategory extends Category {
  missing: boolean;
}

export interface ExitPassedProps {
  exit: Exit;
  categories: Category[];
  node: FlowNode;
  showDragHelper: boolean;
  plumberMakeSource: (id: string) => void;
  plumberRemove: (id: string) => void;
  plumberConnectExit: (node: FlowNode, exit: Exit) => void;
  plumberUpdateClass: (
    node: FlowNode,
    exit: Exit,
    className: string,
    confirmDelete: boolean
  ) => void;
}

export interface ExitStoreProps {
  translating: boolean;
  dragging: boolean;
  language: Asset;
  localization: LocalizationMap;
  disconnectExit: DisconnectExit;
  segmentCount: number;
  recentMessages: RecentMessage[];
}

export type ExitProps = ExitPassedProps & ExitStoreProps;

export interface ExitState {
  confirmDelete: boolean;
  recentMessages: RecentMessage[];
  fetchingRecentMessages: boolean;
  showDragHelper: boolean;
}

const cx: any = classNames.bind(styles);
export class ExitComp extends React.PureComponent<ExitProps, ExitState> {
  private timeout: number;
  private hideDragHelper: number;
  private pendingMessageFetch: Cancel = {};
  private ele: HTMLDivElement;

  constructor(props: ExitProps) {
    super(props);

    this.state = {
      confirmDelete: false,
      recentMessages: null,
      fetchingRecentMessages: false,
      showDragHelper: props.showDragHelper
    };

    bindCallbacks(this, {
      include: [/^on/, /^get/, /^handle/, /^connect/]
    });
  }

  public static contextTypes = {
    config: fakePropType
  };

  public getSourceId(): string {
    return `${this.props.node.uuid}:${this.props.exit.uuid}`;
  }

  public handleDisconnect(): void {
    this.setState({ showDragHelper: false });
  }

  public componentDidMount(): void {
    this.props.plumberMakeSource(this.getSourceId());

    if (this.ele) {
      this.ele.addEventListener('disconnect', this.handleDisconnect);
    }

    if (this.props.exit.destination_uuid) {
      this.connect();
    }
  }

  public componentDidUpdate(prevProps: ExitProps): void {
    if (this.props.exit.destination_uuid !== prevProps.exit.destination_uuid) {
      this.connect();
      if (this.state.confirmDelete) {
        this.setState({ confirmDelete: false });
      }
    }

    if (this.state.showDragHelper && prevProps.showDragHelper && !this.props.showDragHelper) {
      this.setState({ showDragHelper: false });
    }

    this.props.plumberUpdateClass(
      this.props.node,
      this.props.exit,
      'confirm-delete',
      this.state.confirmDelete
    );
  }

  public componentWillUnmount(): void {
    if (this.props.exit.destination_uuid) {
      this.props.plumberRemove(this.getSourceId());
    }

    if (this.ele) {
      this.ele.removeEventListener('disconnect', this.handleDisconnect);
    }
  }

  private handleMouseDown(event: React.MouseEvent<HTMLElement>): void {
    event.preventDefault();
    event.stopPropagation();
  }

  private handleClick(event: React.MouseEvent<HTMLElement>): void {
    if (!this.props.translating) {
      if (this.props.exit.destination_uuid) {
        event.preventDefault();
        event.stopPropagation();
        this.setState(
          {
            confirmDelete: true
          },
          () => {
            this.timeout = window.setTimeout(() => {
              this.setState({
                confirmDelete: false
              });
            }, 2000);
          }
        );
      } else {
        event.preventDefault();
        event.stopPropagation();
        if (!this.state.showDragHelper) {
          this.setState({ showDragHelper: true }, () => {
            if (this.hideDragHelper) {
              window.clearTimeout(this.hideDragHelper);
            }
            this.hideDragHelper = window.setTimeout(() => {
              this.setState({ showDragHelper: false });
            }, 3000);
          });
        }
      }
    }
  }

  private onDisconnect(event: React.MouseEvent<HTMLElement>): void {
    if (this.timeout) {
      window.clearTimeout(this.timeout);
    }

    if (this.hideDragHelper) {
      window.clearTimeout(this.hideDragHelper);
    }

    this.props.disconnectExit(this.props.node.uuid, this.props.exit.uuid);
  }

  private connect(): void {
    this.props.plumberConnectExit(this.props.node, this.props.exit);
  }

  private handleShowRecentMessages(): void {
    if (this.props.recentMessages) {
      this.setState({ recentMessages: this.props.recentMessages });
      return;
    }

    this.setState({ fetchingRecentMessages: true }, () => {
      getRecentMessages(
        this.context.config.endpoints.recents,
        this.props.exit,
        this.pendingMessageFetch
      )
        .then((recentMessages: RecentMessage[]) => {
          this.setState({ recentMessages, fetchingRecentMessages: false });
        })
        .catch(() => {
          // we may have been canceled
        });
    });
  }

  private handleHideRecentMessages(): void {
    if (this.pendingMessageFetch.reject) {
      this.pendingMessageFetch.reject();
      this.pendingMessageFetch = {};
    }

    this.setState({ fetchingRecentMessages: false, recentMessages: null });
  }

  private getSegmentCount(): JSX.Element {
    // Only exits with a destination have activity
    if (this.props.segmentCount > 0) {
      const key = `${this.props.exit.uuid}-label`;
      return (
        <div style={{ position: 'absolute', bottom: '-25px' }}>
          <Counter
            key={key}
            count={this.props.segmentCount}
            containerStyle={styles.activity}
            countStyle={styles.count}
            keepVisible={false}
            onMouseEnter={this.handleShowRecentMessages}
            onMouseLeave={this.handleHideRecentMessages}
          />
        </div>
      );
    }
  }

  public getName(): { name: string; localized?: boolean } {
    if (this.props.translating) {
      let name: string = '';
      let delim: string = '';

      let localized: boolean = false;

      this.props.categories.forEach((category: Category) => {
        const localization = getLocalization(
          category,
          this.props.localization,
          this.props.language
        );

        localized = localized || 'name' in localization.localizedKeys;
        const localizedObject = localization.getObject() as Category;
        name += delim + localizedObject.name;
        delim = ', ';
      });

      return { name, localized };
    } else {
      const names: string[] = [];
      this.props.categories.forEach((cat: Category) => {
        names.push(cat.name);
      });

      return {
        name: names.join(', ')
      };
    }
  }

  private getRecentMessages(): JSX.Element {
    if (this.state.fetchingRecentMessages || this.state.recentMessages !== null) {
      const recentMessages = this.state.recentMessages || [];
      const hasRecents = recentMessages.length !== 0;

      const recentStyles = [styles.recent_messages];

      let title = i18n.t('recent_messages.header', 'Recent Messages');
      if (!hasRecents && !this.state.fetchingRecentMessages) {
        title = i18n.t('recent_messages.header_empty', 'No Recent Messages');
        recentStyles.push(styles.no_recents);
      }

      const canvas = document.getElementById('canvas-container');
      let left = 0;
      let top = 0;

      if (canvas) {
        const canvasBounds = canvas.getBoundingClientRect();
        const canvasOffset = canvasBounds.top + window.scrollY;
        const rect = this.ele.getBoundingClientRect();
        left = rect.left + window.scrollX + 5;
        top = rect.top + window.scrollY - canvasOffset + 30;
      }

      return (
        <Portal id="activity_recent_messages">
          <div className={recentStyles.join(' ')} style={{ position: 'absolute', left, top }}>
            <div className={styles.title}>{title}</div>
            {recentMessages.map((recentMessage: RecentMessage, idx: number) => (
              <div key={'recent_' + idx} className={styles.message}>
                <div className={styles.text}>{recentMessage.text}</div>
                <div className={styles.sent}>{moment.utc(recentMessage.sent).fromNow()}</div>
              </div>
            ))}
            {this.state.recentMessages === null ? (
              <div className={styles.loading}>
                <Loading size={10} units={6} color="#999999" />
              </div>
            ) : null}
          </div>
        </Portal>
      );
    }
    return null;
  }

  public render(): JSX.Element {
    const { name, localized } = this.getName();

    const nameStyle = name ? styles.name : '';
    const connected = this.props.exit.destination_uuid ? ' jtk-connected' : '';
    const dragNodeClasses = cx(styles.endpoint, connected);
    const confirmDelete =
      this.state.confirmDelete && this.props.exit.hasOwnProperty('destination_uuid');
    const confirm: JSX.Element =
      confirmDelete && this.context.config.mutable ? (
        <div
          className={styles.confirm_x + ' fe-x'}
          {...createClickHandler(this.onDisconnect, () => this.props.dragging)}
        />
      ) : null;
    const exitClasses: string = cx({
      [styles.exit]: true,
      'plumb-exit': true,
      [styles.translating]: this.props.translating,
      [styles.unnamed_exit]: name == null,
      [styles.missing_localization]: name && this.props.translating && !localized,
      [styles.confirm_delete]: confirmDelete
    });

    const activity = this.getSegmentCount();
    const recents = this.getRecentMessages();

    const events = this.context.config.mutable
      ? createClickHandler(
          this.handleClick,
          () => {
            return this.props.dragging;
          },
          this.handleMouseDown
        )
      : {};

    return (
      <div className={exitClasses}>
        {name ? <div className={nameStyle}>{name}</div> : null}
        <div
          ref={(ref: HTMLDivElement) => (this.ele = ref)}
          {...events}
          id={`${this.props.node.uuid}:${this.props.exit.uuid}`}
          className={dragNodeClasses}
        >
          {confirm}
        </div>
        {activity}
        {recents}
        {renderIf(this.state.showDragHelper)(<DragHelper />)}
      </div>
    );
  }
}

const mapStateToProps = (
  {
    flowContext: {
      definition: { localization }
    },
    editorState: { translating, language, dragActive, activity }
  }: AppState,
  props: ExitPassedProps
) => {
  // see if we have some passed in (simulated) messages
  let recentMessages: RecentMessage[] = null;
  const key = getExitActivityKey(props.exit);
  if (key in (activity.recentMessages || {})) {
    recentMessages = activity.recentMessages[key];
  }
  const segmentCount = activity.segments[getExitActivityKey(props.exit)] || 0;

  return {
    dragging: dragActive,
    segmentCount,
    translating,
    language,
    localization,
    recentMessages
  };
};

const mapDispatchToProps = (dispatch: DispatchWithState) =>
  bindActionCreators({ disconnectExit }, dispatch);

const ConnectedExit = connect(
  mapStateToProps,
  mapDispatchToProps
)(ExitComp);

export default ConnectedExit;
