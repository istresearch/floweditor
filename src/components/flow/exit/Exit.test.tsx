import * as React from 'react';
import { render, fireEvent, findByText, getByText } from 'test/utils';
import { ExitComp, ExitProps } from './Exit';
import { createUUID } from 'utils';
import { Exit, FlowNode } from 'flowTypes';
import { Spanish } from 'testUtils/assetCreators';

const exit: Exit = {
  uuid: createUUID(),
  destination_uuid: createUUID()
};
const categories = [{ uuid: createUUID(), name: 'Red', exit_uuid: exit.uuid }];

const exitProps: ExitProps = {
  exit,
  categories,
  recentMessages: [],
  node: { uuid: createUUID(), actions: [], exits: [] },
  showDragHelper: false,
  translating: false,
  dragging: false,
  language: null,
  localization: null,
  segmentCount: 1000,
  plumberConnectExit: (node: FlowNode, exit: Exit) => {},
  plumberRemove: jest.fn(),
  plumberMakeSource: jest.fn(),
  plumberUpdateClass: jest.fn(),
  disconnectExit: jest.fn()
};

jest.useFakeTimers();

describe(ExitComp.name, () => {
  it('renders', () => {
    const { baseElement } = render(<ExitComp {...exitProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('shows activity', () => {
    const { baseElement, getByText } = render(
      <>
        <ExitComp
          {...exitProps}
          recentMessages={[
            { text: 'Hi Mom!', sent: new Date().toUTCString() },
            { text: 'Hi Dad!', sent: new Date().toUTCString() }
          ]}
        />
      </>
    );

    // give our portal a chance to mount
    jest.runAllTimers();

    // we have activity but we can't see our recent messages yet
    getByText('1,000');
    expect(baseElement).toMatchSnapshot(baseElement);
  });

  it('shows recent messages on mouse over', () => {
    const { baseElement, getByText, queryAllByText } = render(
      <>
        <div id="activity_recent_messages"></div>

        <ExitComp
          {...exitProps}
          recentMessages={[
            { text: 'Hi Mom!', sent: new Date().toUTCString() },
            { text: 'Hi Dad!', sent: new Date().toUTCString() }
          ]}
        />
      </>
    );

    // give our portal a chance to mount
    jest.runAllTimers();

    // now we need to mouse over our activity to see recent messages
    const activity = getByText('1,000');
    expect(queryAllByText('Recent Messages').length).toEqual(0);

    fireEvent.mouseEnter(activity);
    jest.runAllTimers();

    expect(queryAllByText('Recent Messages').length).toEqual(1);
    getByText('Hi Mom!');
    getByText('Hi Dad!');
    expect(baseElement).toMatchSnapshot();
  });

  it('shows missing localization', () => {
    const { baseElement, container } = render(
      <ExitComp {...exitProps} translating={true} language={Spanish} localization={{ spa: {} }} />
    );

    expect(container.querySelector('.missing_localization')).not.toBe(null);
    expect(baseElement).toMatchSnapshot();
  });

  it('shows localization', () => {
    const { baseElement, getByText } = render(
      <ExitComp
        {...exitProps}
        translating={true}
        language={Spanish}
        localization={{ spa: { [categories[0].uuid]: { name: ['Hola!'] } } }}
      />
    );

    getByText('Hola!');
    expect(baseElement).toMatchSnapshot();
  });
});
