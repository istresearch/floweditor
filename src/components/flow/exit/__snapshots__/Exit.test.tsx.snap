// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExitComp renders 1`] = `
<body>
  <div>
    <div
      class="exit plumb-exit"
    >
      <div
        class="name"
      >
        Red
      </div>
      <div
        class="endpoint  jtk-connected"
        id="763e4844-3e1b-407a-a1b5-5fdfcd308b41:b1f332f3-bdd3-4891-aec5-1843a712dbf1"
      />
      <div
        style="position: absolute; bottom: -25px;"
      >
        <div
          class="counter activity  visible"
          data-spec="counter-outter"
        >
          <div
            class="count"
            data-spec="counter-inner"
          >
            1,000
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ExitComp shows activity 1`] = `
Object {
  Symbol(SameObject caches): Object {
    "attributes": NamedNodeMap {},
    "childNodes": NodeList [
      <div>
        <div
          class="exit plumb-exit"
        >
          <div
            class="name"
          >
            Red
          </div>
          <div
            class="endpoint  jtk-connected"
            id="763e4844-3e1b-407a-a1b5-5fdfcd308b41:b1f332f3-bdd3-4891-aec5-1843a712dbf1"
          />
          <div
            style="position: absolute; bottom: -25px;"
          >
            <div
              class="counter activity  visible"
              data-spec="counter-outter"
            >
              <div
                class="count"
                data-spec="counter-inner"
              >
                1,000
              </div>
            </div>
          </div>
        </div>
      </div>,
    ],
  },
}
`;

exports[`ExitComp shows localization 1`] = `
<body>
  <div>
    <div
      class="exit plumb-exit translating"
    >
      <div
        class="name"
      >
        Hola!
      </div>
      <div
        class="endpoint  jtk-connected"
        id="763e4844-3e1b-407a-a1b5-5fdfcd308b41:b1f332f3-bdd3-4891-aec5-1843a712dbf1"
      />
      <div
        style="position: absolute; bottom: -25px;"
      >
        <div
          class="counter activity  visible"
          data-spec="counter-outter"
        >
          <div
            class="count"
            data-spec="counter-inner"
          >
            1,000
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ExitComp shows missing localization 1`] = `
<body>
  <div>
    <div
      class="exit plumb-exit translating missing_localization"
    >
      <div
        class="name"
      >
        Red
      </div>
      <div
        class="endpoint  jtk-connected"
        id="763e4844-3e1b-407a-a1b5-5fdfcd308b41:b1f332f3-bdd3-4891-aec5-1843a712dbf1"
      />
      <div
        style="position: absolute; bottom: -25px;"
      >
        <div
          class="counter activity  visible"
          data-spec="counter-outter"
        >
          <div
            class="count"
            data-spec="counter-inner"
          >
            1,000
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ExitComp shows recent messages on mouse over 1`] = `
<body>
  <div>
    <div
      id="activity_recent_messages"
    >
      <div>
        <div
          class="recent_messages"
          style="position: absolute; left: 0px; top: 0px;"
        >
          <div
            class="title"
          >
            Recent Messages
          </div>
          <div
            class="message"
          >
            <div
              class="text"
            >
              Hi Mom!
            </div>
            <div
              class="sent"
            >
              a few seconds ago
            </div>
          </div>
          <div
            class="message"
          >
            <div
              class="text"
            >
              Hi Dad!
            </div>
            <div
              class="sent"
            >
              a few seconds ago
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="exit plumb-exit"
    >
      <div
        class="name"
      >
        Red
      </div>
      <div
        class="endpoint  jtk-connected"
        id="763e4844-3e1b-407a-a1b5-5fdfcd308b41:b1f332f3-bdd3-4891-aec5-1843a712dbf1"
      />
      <div
        style="position: absolute; bottom: -25px;"
      >
        <div
          class="counter activity  visible"
          data-spec="counter-outter"
        >
          <div
            class="count"
            data-spec="counter-inner"
          >
            1,000
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
